accord-core/src/main/java/accord/messages/Await.java

```diff
@@ -0,0 +1,268 @@
+    private transient ReplyContext replyContext;
+
+    private transient volatile RemoteListeners.Registration registration;
+    private static final AtomicReferenceFieldUpdater<Await, RemoteListeners.Registration> registrationUpdater = AtomicReferenceFieldUpdater.newUpdater(Await.class, RemoteListeners.Registration.class, "registration");
+
+    private transient volatile int synchronouslyWaitingOn;
+    private static final AtomicIntegerFieldUpdater<Await> synchronouslyWaitingOnUpdater = AtomicIntegerFieldUpdater.newUpdater(Await.class, "synchronouslyWaitingOn");
+
+    public Await(Id to, Topologies topologies, TxnId txnId, Participants<?> participants, BlockedUntil blockedUntil)
+    {
+        this(to, topologies, txnId, participants, blockedUntil, Integer.MIN_VALUE);
+    }
+
+    public Await(Id to, Topologies topologies, TxnId txnId, Participants<?> participants, BlockedUntil blockedUntil, int callbackId)
+    {
+        this.txnId = txnId;
+        this.callbackId = callbackId;
+        this.scope = computeScope(to, topologies, participants);
+        this.blockedUntil = blockedUntil;
+    }
+
+    public Await(Id to, Topology topology, TxnId txnId, Participants<?> participants, BlockedUntil blockedUntil)
+    {
+        this(to, topology, txnId, participants, blockedUntil, Integer.MIN_VALUE);
+    }
+
+    public Await(Id to, Topology topology, TxnId txnId, Participants<?> participants, BlockedUntil blockedUntil, int callbackId)
+    {
+        this.txnId = txnId;
+        this.scope = participants.slice(topology.rangesForNode(to));
+        this.blockedUntil = blockedUntil;
+        this.callbackId = callbackId;
+    }
+
+    private Await(TxnId txnId, Participants<?> scope, BlockedUntil blockedUntil, int callbackId)
+    {
+        this.txnId = txnId;
+        this.scope = scope;
+        this.blockedUntil = blockedUntil;
+        this.callbackId = callbackId;
+    }
+
+    @Override
+    public void process(Node node, Id replyToNode, ReplyContext replyContext)
+    {
+        this.node = node;
+        this.replyTo = replyToNode;
+        this.replyContext = replyContext;
+        node.mapReduceConsumeLocal(this, scope, txnId.epoch(), txnId.epoch(), this);
+    }
+
+    @Override
+    public Void apply(SafeCommandStore safeStore)
+    {
+        SafeCommand safeCommand = safeStore.get(txnId, txnId, scope);
+        Command command = safeCommand.current();
+        if (command.saveStatus().compareTo(blockedUntil.minSaveStatus) >= 0)
+            return null;
+
+        // TODO (desired): consider merging these methods?
+        command = Commands.updateRouteOrParticipants(safeStore, safeCommand, scope);
+
+        if (callbackId >= 0)
+        {
+            RemoteListeners.Registration registered = registration;
+            if (registered == null)
+            {
+                registered = node.remoteListeners().register(txnId, blockedUntil.minSaveStatus, blockedUntil.remoteDurability, replyTo, callbackId);
```

Comment:
that's fine I suppose, although we hope not to have to do this work at all as we hope it will be ready to reply. No major strong feelings about this as it isn't an onerous object to create unnecessarily, just pointing out the rationale